### PR TITLE
Peer Persistence based Replication

### PR DESCRIPTION
**Pull request by imran-ansari**
@wdurairaj , @hpe-storage - Please review the code.

### CODE REVIEW

**Comment by wdurairaj** [comment]
@imran-ansari 

One comment on this section definition for replication

```
replication_device = backend_id:<Array-Name>,
                     replication_mode:<synchronous|asynchronous|streaming>,
                     cpg_map:<Source-CPG>:<Target-CPG>,
                     snap_cpg_map:<Source-Snap-CPG>:<Target-Snap-CPG>
                     hpe3par_api_url:https://<IP>:8080/api/v1,
                     hpe3par_username:<3PAR-Username>,
                     hpe3par_password:<3PAR-Password>,
                     san_ip:<IP>,
                     san_login:<3PAR-SAN-Username>,
                     san_password:<3PAR-SAN-Password>

```

Since we are supporting multi-backend now, backend-id: <array_name> can be `backend-id: <backend-name>` , and the other details like 
```
 hpe3par_api_url:https://<IP>:8080/api/v1,
                     hpe3par_username:<3PAR-Username>,
                     hpe3par_password:<3PAR-Password>,
                     san_ip:<IP>,
                     san_login:<3PAR-SAN-Username>,
                     san_password:<3PAR-SAN-Password>
```
can be read from the corresponding backend section defined. Just a thought. You may want to do this on another pull request preferably.

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 1]
```diff
@@ -1366,3 +1323,155 @@ def get_snapshots_by_vol(self, vol_id, snap_cpg):
...
+        rcg_targets = []
+        rcg_target = {'targetName': self.tgt_bkend_config.backend_id,
+                      'secVolumeName': bkend_vol_name}
+        rcg_targets.append(rcg_target)
+        optional = {'volumeAutoCreation': True}
+        try:
+            rcg = self.get_rcg(rcg_name)
+            original_rcg_state = rcg['targets'][0]['state']
+            if original_rcg_state == self.RCG_STARTED:
+                self.client.stopRemoteCopy(rcg_name)
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_common.py]: Sure

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 2]
```diff
@@ -1366,3 +1323,155 @@ def get_snapshots_by_vol(self, vol_id, snap_cpg):
...
+        except Exception as ex:
+            msg = (_("There was an error fetching the remote copy "
+                     "group after its creation: %s.") % six.text_type(ex))
+            LOG.error(msg)
+            raise exception.HPERemoteCopyGroupBackendAPIException(data=msg)
+
+    def delete_rcg(self, **kwargs):
+        rcg_name = kwargs['rcg_name']
+        try:
+            self.client.removeRemoteCopyGroup(rcg_name)
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_common.py]: Sure

File: hpedockerplugin/hpe/hpe_3par_fc.py [hunk 3]
```diff
@@ -62,16 +57,22 @@ class HPE3PARFCDriver(object):
...
+                 tgt_bkend_config=None):
         self.configuration = hpe3parconfig
-        self.configuration.append_config_values(hpecommon.hpe3par_opts)
 
-        self.hpe3parconfig.append_config_values(san_driver.san_opts)
-        self.hpe3parconfig.append_config_values(san_driver.volume_opts)
+        # self.configuration.append_config_values(hpecommon.hpe3par_opts)
+        # self.hpe3parconfig.append_config_values(san_driver.san_opts)
+        # self.hpe3parconfig.append_config_values(san_driver.volume_opts)
+
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_fc.py]: Okay

**Pull request merged**